ext/openssl/openssl.c

```diff
@@ -1399,6 +1399,48 @@ static X509 *php_openssl_x509_from_str(zend_string *cert_str) {
 			return NULL;
 		}
 		cert = PEM_read_bio_X509(in, NULL, NULL, NULL);
+	} else if (ZSTR_LEN(cert_str) > 7 && memcmp(ZSTR_VAL(cert_str), "pkcs11:", sizeof("pkcs11:") - 1) == 0) {
+		char *verbose = NULL;
+		ENGINE *engine;
+	        struct {
+			const char *s_slot_cert_id;
+			X509 *cert;
+		} parms = {
+			.s_slot_cert_id = ZSTR_VAL(cert_str),
+			.cert = NULL,
+		};
+		int force_login = 0;
+
+		engine = ENGINE_by_id("pkcs11");
+		if (engine == NULL) {
+			php_openssl_store_errors();
+			return NULL;
+		}
+		verbose = getenv("OPENSSL_ENGINE_VERBOSE");
+		if (verbose) {
+			if (!ENGINE_ctrl_cmd_string(engine, "VERBOSE", NULL, 0)) {
+				ENGINE_free(engine);
+				php_openssl_store_errors();
+				return NULL;
+			}
+		}
+		if (!ENGINE_init(engine)) {
+			ENGINE_free(engine);
+			php_openssl_store_errors();
+			return NULL;
+		}
+		if (!ENGINE_ctrl_cmd(engine, "LOAD_CERT_CTRL", 0, &parms, NULL, force_login)) {
+			ENGINE_free(engine);
```

Comment:
It should probably do `ENGINE_finish` as well, right?

Reply:
According to openssl/apps/apps.c ; I do not see calls to `ENGINE_finish()` when `ENGINE_ctrl_cmd()` fails, do I miss something ? Why do you believe it should be called ? Which reference are you using please ?

Reply:
It's based on the [ENGINE_init](https://www.openssl.org/docs/man1.1.1/man3/ENGINE_init.html) docs. Specifically:
```
To obtain a functional reference from an existing structural reference, call the ENGINE_init() function. This returns zero if the ENGINE was not already operational and couldn't be successfully initialised (e.g. lack of system drivers, no special hardware attached, etc), otherwise it will return nonzero to indicate that the ENGINE is now operational and will have allocated a new functional reference to the ENGINE. All functional references are released by calling ENGINE_finish() (which removes the implicit structural reference as well).
```
IIUIC it means that after `ENGINE_init`, there should be `ENGINE_finish` to release the functional reference. I have checked the code though but expecting that docs are correct.

Reply:
done: https://github.com/php/php-src/pull/6860/commits/de9cc93182cdccf4f81e52880d42ae669f9a7de6, thanks for your reviews.